sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2541,7 +2542,20 @@ class Dataset[T] private[sql](
   def dropDuplicates(colNames: Seq[String]): Dataset[T] = withTypedPlan {
     val resolver = sparkSession.sessionState.analyzer.resolver
     val allColumns = queryExecution.analyzed.output
-    val groupCols = colNames.distinct.flatMap { (colName: String) =>
+    // SPARK-31990: We must preserve the input order of `colNames` because of the compatibility
+    // issue (the Streaming's state store depends on the `groupCols` order).
+    val orderPreservingDistinctColNames = {
+      val nameSeen = mutable.Set[String]()
```

Comment:
This looks logically the same with Scala's `distinct` implementation. Is there a change?

Reply:
Ah, I see and it looks the same: https://github.com/scala/scala/blob/2.12.x/src/library/scala/collection/SeqLike.scala#L504-L523 But, (I'm not a Scala compiler expert though),  the `distinct` scala API makes sure an input order is preserved? I read [the @HeartSaVioR comment ](https://issues.apache.org/jira/browse/SPARK-31990?focusedCommentId=17135286&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17135286) in the Jira, then I'm not sure about that. So, I wrote this method to preserve the order explicitly for safeguard.

Reply:
I think Scala wouldn't change the implementation unless there's strong reason to do so, but in Scala 2.13 they removed the orderness guarantee on the method doc in `distinct`. Till Scala 2.12 it clearly explains the return value as preserving the "first occurrence", and the explanation no longer exists in Scala 2.13.

Reply:
So consider this as manual implementation of distinct so that we don't even get affected by Scala changes.

Reply:
Oh I didn't see @maropu 's comment while I'm commenting. ;) Thanks for explaining.

Reply:
The reported issue claims that Scala `distinct` function was not enough. That's the reason why I asked that `What is the difference to fix Spark issue`.

As @HeartSaVioR 's commented (https://github.com/apache/spark/pull/28830#discussion_r439904302), we need a different code.

Reply:
Ah, I see. Yea, I'll update the code based on `toSeq.toSeq`.

Reply:
How about simply revert this line to https://github.com/apache/spark/pull/28062/files#diff-7a46f10c3cedbf013cf255564d9483cdL2458, use the original implementation of `toSet`.
Yes, the `toSet.toSeq` might incompatible during to Scala version, but I think the current fix should just keep the original order. How to detect the order changing and have solid validation should be the work of SPARK-31894 and SPARK-27237. WDYT?